### ISSUE TITLE
readline uses read_nonblock

### ISSUE THREAD

**Issue by cydonian** [excon/excon#471]
I'm using windows (8.1), cygwin (1.7.34), vagrant (1.7.2) , vagrant-openstack-plugin (0.11.1), which then uses excon 0.44.1.

Until yesterday I was using vagrant 1.6.5 and vagrant-openstack-plugin (0.9.0) installed on 24-Nov-2014. At that time excon 0.41.0 was used.

After upgrading my vagrant configuration, excon 0.44.1 was installed. Now I get an error like this

```
$ vagrant up --provider=openstack
C:/Users/cydonian/.vagrant.d/gems/gems/excon-0.44.1/lib/excon/socket.rb:49:in `read_nonblock': A non-blocking socket operation could not be completed immediately. - read would block (Errno::EWOULDBLOCK) (Excon::Errors::SocketError)
        from C:/Users/cydonian/.vagrant.d/gems/gems/excon-0.44.1/lib/excon/socket.rb:49:in `readline'
        from C:/Users/cydonian/.vagrant.d/gems/gems/excon-0.44.1/lib/excon/response.rb:47:in `parse'
        from C:/Users/cydonian/.vagrant.d/gems/gems/excon-0.44.1/lib/excon/middlewares/response_parser.rb:6:in `response_call'
        from C:/Users/cydonian/.vagrant.d/gems/gems/excon-0.44.1/lib/excon/connection.rb:372:in `response'
        from C:/Users/cydonian/.vagrant.d/gems/gems/excon-0.44.1/lib/excon/connection.rb:236:in `request'
        from C:/Users/cydonian/.vagrant.d/gems/gems/fog-core-1.28.0/lib/fog/core/connection.rb:63:in `request'
        from C:/Users/cydonian/.vagrant.d/gems/gems/fog-1.27.0/lib/fog/openstack/core.rb:202:in `retrieve_tokens_v2'
        from C:/Users/cydonian/.vagrant.d/gems/gems/fog-1.27.0/lib/fog/openstack/core.rb:97:in `authenticate_v2'
        from C:/Users/cydonian/.vagrant.d/gems/gems/fog-1.27.0/lib/fog/openstack/compute.rb:395:in `authenticate'
        from C:/Users/cydonian/.vagrant.d/gems/gems/fog-1.27.0/lib/fog/openstack/compute.rb:320:in `initialize'
        from C:/Users/cydonian/.vagrant.d/gems/gems/fog-core-1.28.0/lib/fog/core/service.rb:115:in `new'
        from C:/Users/cydonian/.vagrant.d/gems/gems/fog-core-1.28.0/lib/fog/core/service.rb:115:in `new'
        from C:/Users/cydonian/.vagrant.d/gems/gems/fog-core-1.28.0/lib/fog/compute.rb:64:in `new'
        from C:/Users/cydonian/.vagrant.d/gems/gems/vagrant-openstack-plugin-0.11.1/lib/vagrant-openstack-plugin/action/connect_openstack.rb:46:in `call'
<snip>
```

It appears that `excon::socket.readline` uses `read_nonblock` which is throwing an exception. 

It is not clear to me whether this is a bug in excon or an incompatibility with the way vagrant-openstack-plugin is using the fog+excon.

Should `readline` be non-blocking? Or should blocking behaviour be optional in `readline`? Does that even make sense? Should users of excon expect `readline` to not block.

I don't know enough about ruby, these gems or vagrant internals to actually do anything sensible (like pick different gem versions inside vagrant or hack the other gems to not use `excon::socket.readline`). So I hacked my excon to change this:

```
diff --git a/lib/excon/socket.rb b/lib/excon/socket.rb
index a0b0900..48f73a4 100644
--- a/lib/excon/socket.rb
+++ b/lib/excon/socket.rb
@@ -42,7 +42,11 @@ module Excon
       return legacy_readline if RUBY_VERSION.to_f <= 1.8_7
       begin
         buffer = ''
-        buffer << @socket.read_nonblock(1) while buffer[-1] != "\n"
+      if @nonblock
+        buffer << read_nonblock(1) while buffer[-1] != "\n"
+      else
+        buffer << read_block(1) while buffer[-1] != "\n"
+      end
         buffer
       rescue Errno::EAGAIN
         if timeout_reached('read')
```

Sure that "works", but I'm not sure it's "Right".

See also https://github.com/cloudbau/vagrant-openstack-plugin/issues/105

**Comment by geemus** [comment]
@starbelly I think this might be related to the changes we were working on recently. Could you take a peek? Looks like we are missing a rescue on one of the EWOULDBLOCKs somewhere probably, a bit more stacktrace/evidence over on #440 also. Thanks!

**Comment by starbelly** [comment]
I added EWOULDBLOCK and IO:WaitReadble to the catch block on line 47 ( https://github.com/starbelly/excon/commit/c513a0cea3bd75d3935f46efa5ed33cdb39fd8e2 ). I'd like to see if someone with the same or similar environment could test this before a PR is done. 

 @cydonian could you test this by checking out my would_block branch or by modifying lib/excon/socket.rb to match up with the change I just made?

**Comment by cydonian** [comment]
Thanks starbelly. My testing looks good. I did this:
* undid my hack
* tested to make sure it failed. Pass!
* added your rescue fix
* tested to make sure it succeed. Pass!

(still hacking files in place because I'm clueless :-\ However I'm confident in my test scenario and that I have tested the actual fix)

Many thanks

**Comment by starbelly** [comment]
@cydonian thanks so much. Wasn't sure where I was going to find a similar environment. 
@geemus PR coming up.

**Comment by geemus** [comment]
Thanks!

**Comment by geemus** [comment]
Fixed/released in 0.44.2
